feat : Offline Migrate to CMP

Aditya3815 opened the pull request:
https://github.com/user-attachments/assets/6038c1bc-80f6-4d9a-8bbc-3f9b3494d887

Fixes - [Jira-#MIFOSAC-424](https://mifosforge.jira.com/browse/MIFOSAC-424)

Didn't create a Jira ticket, click [here](https://mifosforge.jira.com/jira/software/c/projects/MIFOSAC/issues/) to create new.

Please Add Screenshots If there are any UI changes.

| Before                                     | After                                  |
|--------------------------------------------|----------------------------------------|
|                                            |                                        |

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the static analysis check `./gradlew check` or `ci-prepush.sh` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them.

Aditya3815 committed:
Feature:offline to cmp draft

Aditya3815 committed:
Merge branch 'kmp-impl' into offline-to-cmp

Aditya3815 committed:
Feature:offline to cmp second draft

Aditya3815 committed:
Merge branch 'kmp-impl' into offline-to-cmp

Aditya3815 committed:
Merge branch 'refs/heads/kmp-impl' into offline-to-cmp

Aditya3815 committed:
Feature:offline to cmp final commit

Aditya3815 committed:
Merge branch 'kmp-impl' into offline-to-cmp

# Conflicts:
#	cmp-navigation/src/commonMain/kotlin/cmp/navigation/di/KoinModules.kt
#	cmp-navigation/src/commonMain/kotlin/cmp/navigation/navigation/FeatureNavHost.kt
#	cmp-navigation/src/commonMain/kotlin/cmp/navigation/navigation/HomeDestinationsScreen.kt

Aditya3815 committed:
Merge remote-tracking branch 'origin/offline-to-cmp' into offline-to-cmp

Aditya3815 committed:
Feature:offline to cmp final commit

revanthkumarJ commented on cmp-navigation/src/commonMain/kotlin/cmp/navigation/di/KoinModules.kt:
```diff
@@ -62,14 +63,15 @@ object KoinModules {
             SplashModule,
             SettingsModule,
+//            SplashModule,
         )
```
splash is already there remove this

revanthkumarJ commented on feature/offline/build.gradle.kts:
```diff
@@ -8,18 +8,26 @@
-    alias(libs.plugins.mifos.android.library.jacoco)
+//    alias(libs.plugins.mifos.android.feature)
+//    alias(libs.plugins.mifos.android.library.compose)
+//    alias(libs.plugins.mifos.android.library.jacoco)
```
remove

revanthkumarJ commented on feature/offline/build.gradle.kts:
```diff
@@ -8,18 +8,26 @@
-
-    //material
-    implementation (libs.androidx.material.v168)
+//    implementation (libs.androidx.material.v168)
```
remove this block

revanthkumarJ commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsScreen.kt:
```diff
@@ -221,12 +217,11 @@ private fun PayloadField(
+//    Log.d("C", context.packageName)
 //    if (Network.isOnline(context)) {
     syncCenterPayloads()
 //    } else {
```
i think we have to implement may be it is commented temporarily during migartion.
@itsPronay i think that is commented temporarily right

itsPronay replied:
@revanthkumarJ you mean `    Log.d("C", context.packageName)`? 
However syncing will not properly work since room has not been migrated property, we need to look into it after the KMP migration

HekmatullahAmin replied:
In `SharedApp.kt` it is already handled to show snackbar that the user is not connected to internet. But in here it can be a good option when someone clicks to show them that they are not connected to internat and in `else` branch they can use snackbar. 
so implementing it is not a heavy task i believe there is a repository where we can listne to isOffline and passed here in screen.

Aditya3815 replied:
Yes @itsPronay that's what I am talking about like it is implemented or not ? Anyways we have to implement it later right?

HekmatullahAmin replied:
@Aditya3815 this feature showing snackbar if not connected is nothing to do with `syncCenterPayloads()`. If this `syncCenterPayloads()` is not working that is fine later when they fix it this will automatically work. 
why to do your job twice and maybe even forget to come again and do it.

biplab1 replied:
@Aditya3815 Please remove the commented-out `if (Network.isOnline(context))` (lines 248-257) block, as the network check and snackbar display are already handled in the `IconButton` click listener (lines 106-114). The `checkNetworkConnectionAndSync` function is correctly designed to receive `isOnline` and `onShowOfflineMessage` as parameters.

Aditya3815 replied:
okay biplab brother

Aditya3815 replied:
Resolved

HekmatullahAmin commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/dashboard/OfflineDashboardViewModel.kt:
```diff
@@ -109,32 +92,46 @@ class OfflineDashboardViewModel(
             _offlineDashboardUiState.value = OfflineDashboardUiState.SyncUiState(updatedList)
         }
     }
+    fun <T> handleDataState(
```
make this function private as it is used only in this view model

Aditya3815 replied:
Resolved

HekmatullahAmin commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/dashboard/OfflineDashboardViewModel.kt:
```diff
@@ -109,32 +92,46 @@ class OfflineDashboardViewModel(
+            flow.collect { state ->
+                when (state) {
+                    is DataState.Success -> setCountOfSyncData(type, state.data.size)
+                    is DataState.Error -> setError(type, state.message)
```
- I see you created a reusable function, which is good. But `Avoid duplicate ViewModel scope launches` currently, both setCountOfSyncData() and setError() launch new coroutines.
If called from within handleDataState, which is already running in viewModelScope, this is a bit redundant.

- so remove viewmodelScope.launch from them

revanthkumarJ replied:
@Aditya3815 have you done it

Aditya3815 replied:
Resolved

HekmatullahAmin commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsViewModel.kt:
```diff
@@ -100,7 +117,7 @@ class SyncCenterPayloadsViewModel(
                 }.collect {
                     centerSyncIndex = 0
-                    mCenterPayloads = it.toMutableList()
+                    mCenterPayloads = it as MutableList<CenterPayloadEntity>
```
this is unchecked cast use it.data

Aditya3815 replied:
Assignment type mismatch: actual type is 'List<CenterPayloadEntity>?', but 'MutableList<CenterPayloadEntity>' was expected.

HekmatullahAmin replied:
what i am saying is that instead of casting it like above use it.data?.toMutableList or whatever

Aditya3815 replied:
Is that Like we have to implement like .collect {result -> when(result){ is ....}

HekmatullahAmin replied:
no directly use. `it.date` returns `List<CenterPayloadEntity>` and then you can change it mutablelist

Aditya3815 replied:
result.data?.let { mCenterPayloads = it.toMutableList() } 
I used this like this. could you please confirm me

Aditya3815 replied:
it ask me to wrap with let variable

HekmatullahAmin commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncClientPayloads/SyncClientPayloadsViewModel.kt:
```diff
@@ -104,14 +120,29 @@ class SyncClientPayloadsViewModel(
-                        syncClientPayload()
+                    when (clientPayloads) {
+                        is DataState.Success -> {
+                            val list = clientPayloads.data
```
in place where you don't use the variable elsewhere then it is not needed to create one. just directly do it in if block

Aditya3815 replied:
mClientSyncIndex = 0. this one is already there before

HekmatullahAmin commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncLoanRepaymentTransaction/SyncLoanRepaymentTransactionScreen.kt:
```diff
@@ -307,12 +301,11 @@ private fun EmptyLoanRepaymentsScreen(
     syncLoanRepaymentTransactions: () -> Unit,
 ) {
-    Log.d("C", context.packageName)
+//    Log.d("C", context.packageName)
```
Same here ask Mr. Rajan in stand up call if he approve implement these

Aditya3815 replied:
Pronay said that Room is not migrated properly we have to migrate that then we have to implement this feature so for now I think migration is enough.

HekmatullahAmin replied:
I already mentioned why it would be good now, have a look at the top discussion

HekmatullahAmin commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncSavingsAccountTransaction/SyncSavingsAccountTransactionViewModel.kt:
```diff
@@ -254,7 +290,7 @@ class SyncSavingsAccountTransactionViewModel(
             SyncSavingsAccountTransactionUiState.Loading
         repository.processTransaction(
-            type,
+            type?:"",
```
- handle it properly by doing it you saying if the savings account type is not defined you pass empty string. which is not correct for an account type. 
- be careful when you are using `!!` and passing default values for making api request.

Aditya3815 replied:
Yes you are right @HekmatullahAmin this could result in incorrect or invalid API calls. 
So can I do like this ? 
 require(!type.isNullOrBlank()){"Account type must not be null or blank"}
 requireNotNull(request) {"Request must not be null"}

HekmatullahAmin commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/dashboard/OfflineDashboardScreen.kt:
```diff
@@ -122,12 +127,20 @@ internal fun OfflineDashboardScreen(
+
+                        val firstError = syncState.list.firstOrNull { it.errorMsg != null }?.errorMsg
+
+                        LaunchedEffect(firstError) {
```
Right now you are showing `snackbar` both globally and per item. is it needed? 
avoid duplicate snackbars which one is needed only use that

Aditya3815 replied:
Here I am using the Globally not per item. yes it is needed

Aditya3815 replied:
I think there is no duplicacy. can you check once.

HekmatullahAmin replied:
- on top you scan the list and take the first item where errorMsg is not null. and then show it in snackbar.
- and later in `lazycolumn` for each item we are already will show snackbar if the error message is not empty

HekmatullahAmin replied:
that's why it is redundant

Aditya3815 replied:
Ohh I see, I am sorry brother for the confusion. Get confused with other files

Aditya3815 replied:
I updated and removed the Launched effect code section for the global approach

Aditya3815 committed:
Feature:offline to cmp latest commit

Aditya3815 committed:
Feature:offline to cmp latest second commit

niyajali committed:
Merge branch 'kmp-impl' into offline-to-cmp

niyajali commented:
@Aditya3815 @revanthkumarJ @HekmatullahAmin @itsPronay @biplab1 guys review this PR

niyajali commented:
@Aditya3815 resolve merge conflicts

Aditya3815 committed:
Merge branch 'kmp-impl' into offline-to-cmp

revanthkumarJ commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsScreen.kt:
```diff
@@ -221,14 +234,20 @@ private fun PayloadField(
+        onShowOfflineMessage()
+    }
+//    Log.d("C", context.packageName)
 //    if (Network.isOnline(context)) {
```
remove these comments

Aditya3815 replied:
Resolved

revanthkumarJ commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsScreen.kt:
```diff
@@ -272,7 +307,7 @@ val sampleCenterPayloads = List(5) { index ->
+@DevicePreview
 @Composable
 private fun CenterPayloadItemPreview() {
 //    val sampleCenterPayload = CenterPayload().apply {
```
here also

Aditya3815 replied:
Resolved @revanthkumarJ

revanthkumarJ replied:
the comments are still there

revanthkumarJ commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncClientPayloads/SyncClientPayloadsScreen.kt:
```diff
@@ -277,25 +291,31 @@ private fun ErrorStateScreen(
+//    syncClientPayloads()
 //    } else {
 //        Toast.makeText(
 //            context,
```
here also

Aditya3815 replied:
Resolved

revanthkumarJ reviewed (changes requested):
@Aditya3815 you have commented things at many places remove them if you have implemented that toast msgs in Snackbar

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsScreen.kt:
```diff
@@ -10,50 +10,52 @@
 package com.mifos.feature.offline.syncCenterPayloads
 
 // import com.mifos.core.common.utils.Network
```
Please remove this commented import statement

Aditya3815 replied:
Resolved @biplab1

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsScreen.kt:
```diff
@@ -87,37 +89,53 @@ internal fun SyncCenterPayloadsScreen(
             modifier = Modifier
                 .padding(paddingValues)
-                .pullRefresh(pullRefreshState),
+                .fillMaxSize(),
```
Instead of `CircularProgressIndicator(modifier = Modifier.align(Alignment.Center))`,  you can use `MifosCircularProgress()` for consistency across all screens

Aditya3815 replied:
it was there before I have not implemented

revanthkumarJ replied:
yes as we have added ui components while migaration before all components are used directtly

revanthkumarJ replied:
@biplab1 is saying for consistency between all screens replace CircularProgressIndicator with MifosCircularProgress, eventhough you didn't added it

Aditya3815 replied:
Okay @revanthkumarJ I am updating my PR with all the changes

Aditya3815 replied:
Added MifosCircularProgress()

Resolved

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsScreen.kt:
```diff
@@ -221,14 +234,20 @@ private fun PayloadField(
 }
 
-@RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
+// @RequiresPermission(Manifest.permission.ACCESS_NETWORK_STATE)
```
Please remove the commented line no. `237`. Also, in line no. `226`, if you're using the Material theme's typography, there's no need to explicitly set `fontWeight`—it's already defined implicitly within the typography style itself.

Aditya3815 replied:
Okay biplab. Thanks

Aditya3815 replied:
Resolved @biplab1

revanthkumarJ replied:
have you removed it it is still showing to me

biplab1 commented:
@Aditya3815 I’ve left some comments — please address them. There are also pending change requests from other reviewers that haven’t been addressed yet. Kindly resolve those as well.

Aditya3815 committed:
Merge branch 'kmp-impl' into offline-to-cmp

Aditya3815 committed:
feat : Offline Migrated to CMP

Aditya3815 commented:
> @Aditya3815 also change the pr title from Feature:offline to cmp draft to feat : Offline Migrated to CMP like that

Changed to  feat : Offline Migrated to CMP

Aditya3815 commented:
> @Aditya3815 resolve the static checks

Static checks resolved

Aditya3815 commented:
> @Aditya3815 resolve merge conflicts

Merge conflict Resolved

revanthkumarJ commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncGroupPayloads/SyncGroupPayloadsScreenRoute.kt:
```diff
@@ -249,33 +264,14 @@ private fun GroupPayloadField(
+    }
 }
-
-// @Preview
```
i think you have to add this preview also

itsPronay commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/dashboard/OfflineDashboardScreen.kt:
```diff
@@ -122,12 +127,20 @@ internal fun OfflineDashboardScreen(
                 )
             } else {
                 when (uiState) {
                     is OfflineDashboardUiState.SyncUiState -> {
+                        val syncState = uiState
+
+                        val firstError = syncState.list.firstOrNull { it.errorMsg != null }?.errorMsg
+
+                        LaunchedEffect(firstError) {
+                            firstError?.let { snackBarHostState.showSnackbar(it) }
+                        }
+
                         LazyColumn {
                             items(uiState.list) { item ->
                                 if (item.count != 0) {
```
@Aditya3815 I am not understanding one thing, in ViewModel you are setting error message for every type right? 
so if every sync function fails there would be 5 errors for 5 offline sync items. 

But here you are only showing the first error!

revanthkumarJ replied:
@Aditya3815 check whether you added all toast messages into snackbar or not once

Aditya3815 committed:
feat : Offline Migrated to CMP minor fix

Aditya3815 committed:
Add navigation route

revanthkumarJ committed:
Merge branch 'kmp-impl' into offline-to-cmp

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsViewModel.kt:
```diff
@@ -60,13 +63,27 @@ class SyncCenterPayloadsViewModel(
 
     fun loadDatabaseCenterPayload() {
         viewModelScope.launch(Dispatchers.IO) {
```
Dispatchers.IO is not required and can be removed.

Aditya3815 replied:
preview provider not supported I think

revanthkumarJ replied:
preview providers won't support but you can give individual state in previews multiple times instead of using preview provider

Aditya3815 replied:
I have added them

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncCenterPayloads/SyncCenterPayloadsViewModel.kt:
```diff
@@ -11,11 +11,13 @@ package com.mifos.feature.offline.syncCenterPayloads
 import com.mifos.core.datastore.UserPreferencesRepository
 import com.mifos.room.entities.center.CenterPayloadEntity
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.IO
```
There is no such kotlinx.coroutines.IO object. Please remove it.

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncClientPayloads/SyncClientPayloadsScreen.kt:
```diff
@@ -328,61 +340,26 @@ private fun SyncClientPayloadsScreenPreview(
     )
 }
 
-// Sample data for previews
-// val sampleClientPayloads = List(5) { index ->
-//    ClientPayload().apply {
-//        firstname = "John$index"
-//        middlename = "Sam$index"
-//        lastname = "Doe$index"
-//        mobileNo = "123456789$index"
-//        externalId = "EXT-$index"
-//        officeId = index
-//        active = index % 2 == 0
-//        activationDate = "2023-07-${15 + index}"
-//        genderId = if (index % 3 == 0) 24 else 22
-//        dateOfBirth = "1990-01-0$index"
-//        errorMessage = if (index % 2 == 0) null else "Error in payload"
-//    }
-// }
```
Can you tell me why the sample data was removed?

Aditya3815 replied:
We can implement it later on also. it was commented and can be implement later on

revanthkumarJ replied:
i think previews can be implemented now like how we done in other prs.
Refer them

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncClientPayloads/SyncClientPayloadsScreen.kt:
```diff
@@ -328,61 +340,26 @@ private fun SyncClientPayloadsScreenPreview(
+    PayloadField(label = "First Name", value = "John")
+}
 
-// @Preview(showBackground = true)
-// @Composable
-// private fun ClientPayloadItemPreview() {
-//    val sampleClientPayload = ClientPayload().apply {
-//        firstname = "John"
-//        middlename = "Michael"
-//        lastname = "Doe"
-//        mobileNo = "1234567890"
-//        externalId = "EXT-001"
-//        genderId = 22
-//        dateOfBirth = "1990-01-01"
-//        officeId = 12
-//        activationDate = "2023-07-15"
-//        active = true
-//        errorMessage = null
-//    }
-//
-//    ClientPayloadItem(payload = sampleClientPayload)
-// }
```
same here, please review

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncClientPayloads/SyncClientPayloadsScreen.kt:
```diff
@@ -328,61 +340,26 @@ private fun SyncClientPayloadsScreenPreview(
+    PayloadField(label = "Mobile No", value = "1234567890")
+}
 
-class PayloadFieldPreviewProvider : PreviewParameterProvider<Pair<String, String>> {
-    override val values = sequenceOf(
-        "First Name" to "John",
-        "Last Name" to "Doe",
-        "Mobile No" to "1234567890",
-        "External ID" to "EXT-001",
-        "Gender" to "Male",
-        "Date of Birth" to "1990-01-01",
-        "Office ID" to "12345",
-        "Activation Date" to "2023-07-15",
-        "Active" to "true",
-    )
```
same here, please review

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncGroupPayloads/SyncGroupPayloadsScreenRoute.kt:
```diff
@@ -249,33 +264,14 @@ private fun GroupPayloadField(
 }
-
-// @Preview
-// @Composable
-// private fun SyncGroupPayloadsScreenPreview() {
-//    SyncGroupPayloadsScreen(
-//        uiState = SyncGroupPayloadsUiState.Success(),
-//        onRefresh = { },
-//        onBackPressed = { },
-//        refreshState = false,
-//        syncGroupPayloads = { },
-//        groupPayloadsList = dummyGroupPayloads,
-//        userStatus = true,
-//    )
-// }
```
same here, please review.

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncGroupPayloads/SyncGroupPayloadsViewModel.kt:
```diff
@@ -70,17 +73,31 @@ class SyncGroupPayloadsViewModel(
+                                if (list.isEmpty()) {
+                                    GroupPayloadEmptyState.NOTHING_TO_SYNC
+                                } else {
+                                    null
```
Can you tell me why a null value has been passed?

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncLoanRepaymentTransaction/SyncLoanRepaymentTransactionScreen.kt:
```diff
@@ -307,43 +315,67 @@ private fun EmptyLoanRepaymentsScreen(
+    }
 }
 
-class SyncLoanRepaymentTransactionUiStateProvider :
-    PreviewParameterProvider<SyncLoanRepaymentTransactionUiState> {
-    override val values = sequenceOf(
-        SyncLoanRepaymentTransactionUiState.ShowProgressbar,
-        SyncLoanRepaymentTransactionUiState.ShowError(R.string.feature_offline_failed_to_load_loanrepayment),
-        SyncLoanRepaymentTransactionUiState.ShowEmptyLoanRepayments("No loan repayments to sync"),
-        SyncLoanRepaymentTransactionUiState.ShowLoanRepaymentTransactions(
-            sampleLoanRepaymentRequests,
-            samplePaymentTypeOptions,
```
same here (preview provider), please review

biplab1 commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncLoanRepaymentTransaction/SyncLoanRepaymentTransactionViewModel.kt:
```diff
@@ -9,15 +9,21 @@
 import com.mifos.room.entities.PaymentTypeOptionEntity
 import com.mifos.room.entities.accounts.loans.LoanRepaymentRequestEntity
 import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.IO
```
kotlinx.coroutines.IO can be removed and if there is Dispatchers.IO in the file, it should be removed too.

Aditya3815 replied:
Removed @biplab1 in latest PR

biplab1 reviewed (commented):
I noticed that the preview providers, sample data for previews have been removed. Can you tell me why? We should keep them since Android Studio preview release already supports compose previews.

itsPronay commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncClientPayloads/SyncClientPayloadsViewModel.kt:
```diff
@@ -106,12 +122,27 @@ class SyncClientPayloadsViewModel(
                     mClientSyncIndex = 0
-                    if (clientPayloads.isNotEmpty()) {
-                        syncClientPayload()
+                    when (clientPayloads) {
```
![image](https://github.com/user-attachments/assets/3b2068d1-edb0-4c2c-a3eb-6b8e3bccd5f6)
@Aditya3815 remove the catch block, we already have it inside dataState

Aditya3815 replied:
Fixed in latest PR

revanthkumarJ reviewed (changes requested):
@Aditya3815 resolve that preview provider thing if you have doubt you can view collection sheet or center pr or any other pr how it is done

itsPronay commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncClientPayloads/SyncClientPayloadsViewModel.kt:
```diff
@@ -66,9 +68,23 @@ class SyncClientPayloadsViewModel(
                 }.collect { clientPayloads ->
-                    mClientPayloads = clientPayloads.toMutableList()
-                    _syncClientPayloadsUiState.value =
-                        SyncClientPayloadsUiState.ShowPayloads(mClientPayloads)
```
Here as well. go through all the viewmodels and remove the catch block wherever we used datastate
![image](https://github.com/user-attachments/assets/24db20fe-ea6e-46ba-9a3d-af0b910b0f5e)

Aditya3815 replied:
Fixed in latest one

itsPronay commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncGroupPayloads/SyncGroupPayloadsScreenRoute.kt:
```diff
@@ -249,33 +264,14 @@ private fun GroupPayloadField(
 }
-
-// @Preview
-// @Composable
-// private fun SyncGroupPayloadsScreenPreview() {
-//    SyncGroupPayloadsScreen(
-//        uiState = SyncGroupPayloadsUiState.Success(),
-//        onRefresh = { },
-//        onBackPressed = { },
-//        refreshState = false,
-//        syncGroupPayloads = { },
-//        groupPayloadsList = dummyGroupPayloads,
-//        userStatus = true,
-//    )
-// }
```
Please don't remove previews

itsPronay replied:
add the dependency called uiToolingPreview in your gradle , take a look at other modules

Aditya3815 replied:
Added uiToolingPreview

itsPronay commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncGroupPayloads/SyncGroupPayloadsViewModel.kt:
```diff
@@ -70,17 +73,31 @@ class SyncGroupPayloadsViewModel(
                 .catch {
                     _syncGroupPayloadsUiState.value =
-                        SyncGroupPayloadsUiState.Error(R.string.feature_offline_error_failed_to_load_groupPayload)
+                        SyncGroupPayloadsUiState.Error(Res.string.feature_offline_error_failed_to_load_groupPayload)
```
remove

itsPronay commented on feature/offline/src/commonMain/kotlin/com/mifos/feature/offline/syncGroupPayloads/SyncGroupPayloadsViewModel.kt:
```diff
@@ -120,19 +137,28 @@ class SyncGroupPayloadsViewModel(
                 .catch {
                     _syncGroupPayloadsUiState.value =
-                        SyncGroupPayloadsUiState.Error(R.string.feature_offline_error_failed_to_update_list)
+                        SyncGroupPayloadsUiState.Error(Res.string.feature_offline_error_failed_to_update_list)
```
here as well

Aditya3815 committed:
Merge branch 'kmp-impl' into offline-to-cmp

Aditya3815 committed:
Minor fix on Viewmodels and add preview. e.g remove Dispatcher.IO
update the ui-tooling preview in libs.versions.toml version 1.7.6 to 1.8.2
Add Preview in build.gradle.kts

itsPronay committed:
fixed issues

itsPronay committed:
fixed issues

revanthkumarJ commented:
@Aditya3815 resolve the static checks

Aditya3815 committed:
Static checks apply

Aditya3815 commented:
@niyajali please merge this PR
